File: envdiff/analyzers/compare.py
from __future__ import annotations

from pathlib import Path

from envdiff.parsers.dotenv import parse_dotenv


def compare_dotenv_files(left: str | Path, right: str | Path) -> dict[str, object]:
    left_result = parse_dotenv(left)
    right_result = parse_dotenv(right)

    left_latest = _latest_by_name(left_result.definitions)
    right_latest = _latest_by_name(right_result.definitions)

    left_names = set(left_latest)
    right_names = set(right_latest)
    shared_names = sorted(left_names & right_names)

    differing = []
    for name in shared_names:
        left_definition = left_latest[name]
        right_definition = right_latest[name]
        if (
            left_definition.normalized_value_kind != right_definition.normalized_value_kind
            or left_definition.value != right_definition.value
        ):
            differing.append(
                {
                    "name": name,
                    "left_value": left_definition.value,
                    "right_value": right_definition.value,
                    "left_kind": left_definition.normalized_value_kind,
                    "right_kind": right_definition.normalized_value_kind,
                }
            )

    return {
        "left_path": str(Path(left)),
        "right_path": str(Path(right)),
        "missing_in_left": sorted(right_names - left_names),
        "missing_in_right": sorted(left_names - right_names),
        "duplicates_in_left": sorted(
            definition.name for definition in left_result.definitions if definition.is_duplicate
        ),
        "duplicates_in_right": sorted(
            definition.name for definition in right_result.definitions if definition.is_duplicate
        ),
        "differing_values": differing,
        "warnings": sorted((*left_result.warnings, *right_result.warnings)),
    }


def _latest_by_name(definitions) -> dict[str, object]:
    latest = {}
    for definition in definitions:
        latest[definition.name] = definition
    return latest

File: internal/analyzers/compare.go
package analyzers

import (
	"github.com/davisbuilds/envdiff/internal/dotenv"
	"github.com/davisbuilds/envdiff/internal/model"
	"github.com/davisbuilds/envdiff/internal/order"
)

func CompareDotenvFiles(left string, right string) (map[string]any, error) {
	leftResult, err := dotenv.Parse(left)
	if err != nil {
		return nil, err
	}
	rightResult, err := dotenv.Parse(right)
	if err != nil {
		return nil, err
	}

	leftLatest := latestByName(leftResult.Definitions)
	rightLatest := latestByName(rightResult.Definitions)
	leftNames := keys(leftLatest)
	rightNames := keys(rightLatest)
	sharedNames := intersection(leftNames, rightNames)

	differing := []map[string]any{}
	for _, name := range sharedNames {
		leftDefinition := leftLatest[name]
		rightDefinition := rightLatest[name]
		if leftDefinition.NormalizedValueKind != rightDefinition.NormalizedValueKind ||
			leftDefinition.Value != rightDefinition.Value {
			differing = append(differing, map[string]any{
				"left_kind":   leftDefinition.NormalizedValueKind,
				"left_value":  leftDefinition.Value,
				"name":        name,
				"right_kind":  rightDefinition.NormalizedValueKind,
				"right_value": rightDefinition.Value,
			})
		}
	}

	warnings := append([]string{}, leftResult.Warnings...)
	warnings = append(warnings, rightResult.Warnings...)

	return map[string]any{
		"differing_values":    differing,
		"duplicates_in_left":  duplicateNames(leftResult.Definitions),
		"duplicates_in_right": duplicateNames(rightResult.Definitions),
		"left_path":           left,
		"missing_in_left":     difference(rightNames, leftNames),
		"missing_in_right":    difference(leftNames, rightNames),
		"right_path":          right,
		"warnings":            order.Strings(warnings),
	}, nil
}

func latestByName(definitions []model.EnvVarDefinition) map[string]model.EnvVarDefinition {
	latest := map[string]model.EnvVarDefinition{}
	for _, definition := range definitions {
		latest[definition.Name] = definition
	}
	return latest
}

func duplicateNames(definitions []model.EnvVarDefinition) []string {
	names := []string{}
	for _, definition := range definitions {
		if definition.IsDuplicate {
			names = append(names, definition.Name)
		}
	}
	return order.Strings(names)
}

func keys(values map[string]model.EnvVarDefinition) []string {
	names := make([]string, 0, len(values))
	for name := range values {
		names = append(names, name)
	}
	return order.Strings(names)
}

func intersection(left []string, right []string) []string {
	rightSet := stringSet(right)
	values := []string{}
	for _, value := range left {
		if _, ok := rightSet[value]; ok {
			values = append(values, value)
		}
	}
	return order.Strings(values)
}

func difference(left []string, right []string) []string {
	rightSet := stringSet(right)
	values := []string{}
	for _, value := range left {
		if _, ok := rightSet[value]; !ok {
			values = append(values, value)
		}
	}
	return order.Strings(values)
}

func stringSet(values []string) map[string]struct{} {
	set := map[string]struct{}{}
	for _, value := range values {
		set[value] = struct{}{}
	}
	return set
}
